src/library/scala/collection/immutable/TreeSet.scala

```diff
@@ -117,16 +117,27 @@ final class TreeSet[A] private (tree: RB.Tree[A, Unit])(implicit val ordering: O
     *  @param elem    a new element to add.
     *  @return        a new $coll containing `elem` and all the elements of this $coll.
     */
-  def incl(elem: A): TreeSet[A] = newSet(RB.update(tree, elem, (), overwrite = false))
+  def incl(elem: A): TreeSet[A] = {
+    val t = RB.update(tree, elem, null, overwrite = false)
+    if(t eq tree) this else newSet(t)
+  }
 
   /** Creates a new `TreeSet` with the entry removed.
     *
     *  @param elem    a new element to add.
     *  @return        a new $coll containing all the elements of this $coll except `elem`.
     */
-  def excl(elem: A): TreeSet[A] =
-    if (!RB.contains(tree, elem)) this
-    else newSet(RB.delete(tree, elem))
+  def excl(elem: A): TreeSet[A] = {
+    val t = RB.delete(tree, elem)
+    if(t eq tree) this else newSet(t)
+  }
+
+  override def concat(that: collection.IterableOnce[A]): TreeSet[A] = {
+    val it = that.iterator
+    var t = tree
+    while (it.hasNext) t = RB.update(t, it.next(), null, overwrite = false)
```

Comment:
Can we do anything clever when `that` is another `TreeSet`? E.g could `TreeSet((0 to 63): _*) ++ TreeSet((64 to 128): _*)` perform some structural sharing?

Reply:
Yes, there are more efficient algorithms for merging two trees or building from an ordered sequence (which should be possible to generalize to inserting into an existing tree from an ordered sequence).

Reply:
Mutable `RBTree`-s will still be beneficial during building, where only have a single `Ordering` instance.

Reply:
I spent the last 2 days implementing the algorithms from https://www.cs.cmu.edu/~guyb/papers/BFS16.pdf. The current version is very close to the paper without any optimizations for our data structures. Bulk `union` (or `concat`) for a few different scenarios combined (overlapping and non-overlapping sets of equal or different sizes) is already up to 20x faster (the larger the sets, the better).

Reply:
Make that 10x when comparing against this PR with the improved single-element concatenation.

Reply:
Wow, that's amazing! Nice work!